src/plugins/data_source/server/plugin.ts

```diff
@@ -53,5 +72,28 @@ export class DataSourcePlugin implements Plugin<DataSourcePluginSetup, DataSourc
     return {};
   }
 
-  public stop() {}
+  public stop() {
+    this.dataSourceService!.stop();
+  }
+
+  private createDataSourceRouteHandlerContext = (
+    dataSourceService: DataSourceServiceSetup,
+    config: DataSourcePluginConfigType,
+    logger: Logger
+  ): IContextProvider<RequestHandler<unknown, unknown, unknown>, 'data_source'> => {
+    return (context, req) => {
+      return {
+        opensearch: {
+          getClient: (dataSourceId: string) => {
+            return new DataSourceClient({
+              dataSourceService,
+              logger,
+              scopedSavedObjectsClient: context.core.savedObjects.client,
+              config,
+            }).asDataSource(dataSourceId);
+          },
+        },
+      };
```

Comment:
let's make DataSourceClient a service style component, and exposes functions to plugin, instead of always creating new objects.